src/reducer.ts

```diff
@@ -0,0 +1,54 @@
+import { Reducer } from "react";
+import produce from "immer";
+
+import { State, Action, ActionType } from "./types";
+
+const reducer: Reducer<State, Action> = (state, action) => {
+  switch (action.type) {
+    case ActionType.INITIALIZE_ITEMS: {
+      const { itemIdentifierKey } = state;
+
+      const {
+        initialSelectedItems = [],
+        initialItems = [],
+      } = action.payload;
+
+      return {
+        itemIdentifierKey,
+        items: initialItems.map((item) => ({
+          ...item,
+          isSelected: initialSelectedItems.includes(item),
+        })),
+      };
+    }
+
+    case ActionType.TOGGLE_SELECTED_STATUS: {
+      const { itemIdentifierKey } = state;
+
+      const { item = {} } = action.payload;
+
+      return produce(state, draftState => {
+        draftState.items.map((itemFound) => {
```

Comment:
Maybe it's not the ideal solution for large lists of data. Every time that the items are toggled, we'll have to map through the whole list.

Reply:
Maybe something like `toggleItems(items)` or support for batching?

```
batch(() => {
  toggleItem(item1)
  // more code
  toggleItem(item2)
})

// and could have support to promises

batch(async () => {
  toggleItem(item1)
  await something()
  // more code
  toggleItem(item2)
  // will performs toggle for item1 and item2 just when finished this promise
})

Reply:
The batch idea is cool, but wouldn't it be asynchronous? For instance, the user will see a delay.

Reply:
I also didn't understand the promise support. Why it would be necessary?

Reply:
I'm really worried here about looping through many items resulting in an event loop block

Reply:
Just some idea, I don't think is necessary, just that if you pass an async function to batch, you will have some problems if you didn't expect to support this, maybe you can send an error in this case (notify the user to don't use async with batch).

Also, you have another option: you can `batch` underhood like `toggleItem` just push to an array the command and you schedule a function with setImmediate to batch the actions.

You have a lot of options, they just change some semantics more internally that you don't expose today (like some `onToggleItem` listener), so isn't a problem today what you choose now

Reply:
Thanks for sharing these alternatives. I'll make some changes in the following weeks and I'll mention here to discuss again the solution.

Reply:
I thought a lot about this and also tested inputs of data with more than 1000 items, which didn't block the main thread. 

For now, there's no need to optimize the code, and would also turn in less maintainable.